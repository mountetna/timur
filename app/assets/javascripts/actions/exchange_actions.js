export const addExchange = (exchange)=>{
  return {
    type: 'ADD_EXCHANGE',
    exchange_name: exchange.exchange_name,
    exchange: exchange
  };x
};

export const removeExchange = (exchange_name)=>{
  return {
    type: 'REMOVE_EXCHANGE',
    exchange_name: exchange_name
  };
};

export class Exchange{
  constructor(dispatch, exchange_name){
    this.dispatch = dispatch;
    this.exchange_name = exchange_name;
  }

  fetch(path, options){
    this.dispatch(
      addExchange({
        exchange_name: this.exchange_name,
        exchange_path: path,
        start_time: new Date()
      })
    );

    let localResponse = (response)=>{
      return new Promise((resolve, reject)=>{
        this.dispatch(removeExchange(this.exchange_name));
        resolve(response);
      })
    };

    return fetch(path, options).then(localResponse);
  }
}
