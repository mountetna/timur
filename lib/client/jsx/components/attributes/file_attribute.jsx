// Framework libraries.
import React, {useState, useEffect, useRef} from 'react';
import {connect} from 'react-redux';
import Modal from 'react-modal';

import {useReduxState} from 'etna-js/hooks/useReduxState';
import {useActionInvoker} from 'etna-js/hooks/useActionInvoker';

import {
  reviseDocument,
  getRevisionTempUrl,
  fetchTempUrl
} from '../../actions/magma_actions';
import ButtonBar from '../button_bar';
import Icon from '../icon';

export const STUB = '::blank';
export const TEMP = '::temp';

// metis:\/\/([^\/]*?)\/([^\/]*?)\/(.*)$
const METIS_PATH_MATCH = (path) =>
  new RegExp(
    '^metis://' +
      // project_name
      '([^/]*?)/' +
      // bucket_name
      '([^/]*?)/' +
      // folder path + filename
      '(.*)$'
  ).test(path);

// We don't have a lot of content, so let's get a smaller Modal
const customStyles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    minWidth: '40%',
    transform: 'translate(-50%, -50%)'
  }
};

const FileValue = ({value}) =>
  !value ? (
    <span className='file-missing'> No file </span>
  ) : value instanceof File ? (
    <span className='file-upload'>
      {' '}
      {value.name} ({value.type}){' '}
    </span>
  ) : value === STUB || value.path === STUB ? (
    <span className='file-blank'> Blank file </span>
  ) : value === TEMP || value.path === TEMP ? (
    <span className='file-blank'> {value.original_files[0].name} </span>
  ) : (
    <a href={value.url}> {value.path} </a>
  );

export default function FileAttribute(props) {
  const invoke = useActionInvoker();
  const [metis, setMetis] = useState(false);
  const [error, setError] = useState(false);
  const fileInputRef = useRef(null);

  const {
    metisSelector,
    closeModal,
    selectMetisFile,
    formatFileRevision,
    setTempRevision
  } = useFileActions(metis, error, setMetis, setError, props);

  let {mode, value, revised_value, document, template, attribute} = props;

  if (mode != 'edit')
    return (
      <div className='attribute file'>
        <FileValue value={value} />
      </div>
    );

  let buttons = [
    {
      type: 'upload',
      click: () => fileInputRef.current.click(),
      title: 'Upload a file from your computer'
    },
    {
      type: 'cloud',
      click: () => setMetis(true),
      title: 'Link a file from Metis'
    },
    {
      type: 'stub',
      click: () =>
        invoke(
          reviseDocument(
            document,
            template,
            attribute,
            formatFileRevision(STUB)
          )
        ),
      title: 'Mark this file as blank'
    },
    {
      type: 'remove',
      click: () =>
        invoke(
          reviseDocument(
            document,
            template,
            attribute,
            formatFileRevision(null)
          )
        ),
      title: 'Remove this file link'
    }
  ];

  return (
    <div className='attribute file'>
      <input
        type='file'
        style={{display: 'none'}}
        ref={fileInputRef}
        onChange={setTempRevision}
      />
      {metis && metisSelector()}
      <ButtonBar className='file-buttons' buttons={buttons} />
      <FileValue value={revised_value} />
    </div>
  );

  // const browserState = useReduxState(
  //   browserStateOf({model_name, record_name, tab_name})
  // );
  // const {view, record, tab, revision, template, can_edit} = browserState;
  // const [mode, setMode] = useState('loading');
  // const loading = !view || !template || !record || !tab_name;
  // const {cancelEdits, approveEdits} = useEditActions(setMode, browserState);
  // const {selectOrShowTab, selectDefaultTab, selectTab, showTab} = useTabActions(
  //   browserState,
  //   setMode
  // );
}

// class FileAttribute extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {metis: false, error: false};
//   }

//   render() {
//     let {
//       mode,
//       value,
//       revised_value,
//       document,
//       template,
//       attribute,
//       reviseDocument
//     } = this.props;
//     let {metis} = this.state;
//     console.log('revised value', revised_value);
//     if (mode != 'edit')
//       return (
//         <div className='attribute file'>
//           <FileValue value={value} />
//         </div>
//       );

//     let buttons = [
//       {
//         type: 'upload',
//         click: () => this.input.click(),
//         title: 'Upload a file from your computer'
//       },
//       {
//         type: 'cloud',
//         click: () => this.setState({metis: true}),
//         title: 'Link a file from Metis'
//       },
//       {
//         type: 'stub',
//         click: () =>
//           reviseDocument(
//             document,
//             template,
//             attribute,
//             this.formatFileRevision(STUB)
//           ),
//         title: 'Mark this file as blank'
//       },
//       {
//         type: 'remove',
//         click: () =>
//           reviseDocument(
//             document,
//             template,
//             attribute,
//             this.formatFileRevision(null)
//           ),
//         title: 'Remove this file link'
//       }
//     ];

//     return (
//       <div className='attribute file'>
//         <input
//           type='file'
//           style={{display: 'none'}}
//           ref={(input) => (this.input = input)}
//           onChange={this.setTempRevision}
//         />
//         {metis && this.metisSelector()}
//         <ButtonBar className='file-buttons' buttons={buttons} />
//         <FileValue value={revised_value} />
//       </div>
//     );
//   }
// }

function useFileActions(metis, error, setMetis, setError, props) {
  const invoke = useActionInvoker();
  const metisPathRef = useRef(null);

  return {
    metisSelector,
    closeModal,
    selectMetisFile,
    formatFileRevision,
    setTempRevision
  };

  function metisSelector() {
    // TODO: would be nice to make this like a folder / file search
    return (
      <Modal
        isOpen={metis}
        contentLabel='Enter Metis path'
        style={customStyles}
        onRequestClose={closeModal}
        appElement={document.querySelector('#root')}
      >
        <div className='attribute modal file-metis-select'>
          <h2>Enter a Metis path</h2>
          <div className='input-box-wrapper'>
            <label htmlFor='metis-path-input'>Metis path:</label>
            <input
              id='metis-path-input'
              className='full_text metis-path-input'
              type='text'
              ref={metisPathRef}
              placeholder='metis://<project>/<bucket>/<file-path>'
            />
            <div className='modal-button-wrapper'>
              <ButtonBar
                className='modal-buttons'
                buttons={[
                  {type: 'check', click: () => selectMetisFile()},
                  {type: 'cancel', click: () => setMetis(false)}
                ]}
              />
              {error ? (
                <p className='file-metis-error'>Invalid Metis path</p>
              ) : (
                ''
              )}
              <div className='modal-buttons pull-right'>
                <Icon
                  className=''
                  icon='question-circle'
                  title='Help'
                  onClick={() =>
                    window.open(
                      'https://mountetna.github.io/timur.html#managing-data-files',
                      '_blank'
                    )
                  }
                />
              </div>
            </div>
          </div>
        </div>
      </Modal>
    );
  }

  function closeModal() {
    setMetis(false);
  }

  function selectMetisFile() {
    let {document, template, attribute} = props;

    const metisPath = metisPathRef.current.value;

    if (!METIS_PATH_MATCH(metisPath)) {
      setError(true);
      return;
    } else {
      setError(false);
      setMetis(false);
    }

    invoke(
      reviseDocument(
        document,
        template,
        attribute,
        formatFileRevision(metisPath)
      )
    );

    metisPathRef.current.value = null;
  }

  function formatFileRevision(value) {
    return {path: value};
  }

  function setTempRevision(e) {
    e.preventDefault();
    console.log('about to call reviseDocument');
    let {document, template, attribute} = props;
    invoke(
      reviseDocument(
        document,
        template,
        attribute,
        Object.assign(formatFileRevision(TEMP), {
          original_files: e.target.files
        })
      )
    );
  }
}
