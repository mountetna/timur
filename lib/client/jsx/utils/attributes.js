const order = (attribute, identifierFirst) => {
  let {attribute_type} = attribute;
  let map = identifierFirst
    ? {
        identifier: 1,
        parent: 2
      }
    : {
        parent: 1,
        identifier: 2
      };

  map = {
    ...map,
    ...{
      collection: 3,
      table: 4,
      child: 5,
      link: 6,
      file: 7,
      image: 8
    }
  };
  return map[attribute_type] || 9;
};

export const sortAttributes = (attributes, identifierFirst = false) => {
  const sortedAttributes = Object.values(attributes).sort(
    (a, b) =>
      order(a, identifierFirst) - order(b, identifierFirst) ||
      a.attribute_name.localeCompare(b.attribute_name)
  );
  let finalAttributes = {};
  sortedAttributes.forEach((attribute, index) => {
    // inject an index_order for default views
    finalAttributes[attribute.attribute_name] = {
      ...attribute,
      index_order: attribute.index_order ? attribute.index_order : index
    };
  });

  return finalAttributes;
};

export const visibleSortedAttributes = (
  attributes,
  identifierFirst = false
) => {
  return Object.values(sortAttributes(attributes, identifierFirst)).filter(
    (attr) => !attr.hidden
  );
};

export const visibleSortedAttributesWithUpdatedAt = (
  attributes,
  identifierFirst = false
) => {
  return Object.values(sortAttributes(attributes, identifierFirst)).filter(
    (attr) => !attr.hidden || attr.attribute_name === 'updated_at'
  );
};
