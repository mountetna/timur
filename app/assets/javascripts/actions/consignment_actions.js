// Class imports.
import { Exchange } from './exchange_actions'

// Module imports.
import { getConsignments } from '../api/timur'
import { showMessages } from './message_actions'

// Add a consignment (the actual JSON data) to the store.
export const addConsignment = (name, consignment)=>{
  return {
    'type': 'ADD_CONSIGNMENT',
    'manifest_name': name,
    'consignment': consignment
  };
};

/*
 * Post a manifest to the query api and send the returned consignment to the
 * store. If things go wrong, show a message with the error.
 */
export const requestConsignments = (manifests, success, error)=>{

  return (dispatch)=>{
    var localSuccess = (response)=>{
      for(var name in response){
        dispatch(addConsignment(name, response[name]));
      }

      if(success != undefined) success(response);
    };

    var localErrorResponse = (response)=>{

      if(response.query){

        var msg = `### For our inquiry:\n\n`;
        msg +=    `\`${JSON.stringify(response.query)}\`\n\n`;
        msg +=    `## this bitter response:\n\n`;
        msg +=    `    ${response.errors}`;
        dispatch(showMessages([msg]));
      }
      else if(response.errors && response.errors.length == 1){

        var msg = `### Our inquest has failed, for this fault:\n\n`;
        msg +=    `    ${response.errors[0]}`;
        dispatch(showMessages([msg]));
      }
      else if(response.errors && response.errors.length > 1){

        var msg = `### Our inquest has failed, for these faults:\n\n`;
        msg +=    `${response.errors.map((error) => `* ${error}`).join('\n')}`;
        dispatch(showMessages([msg]));
      }

      if(error != undefined) error(response);
    };

    var localError = (e)=>{

      e.response.json()
        .then(res);
    };

    var manifest_names = manifests.map(m=>m.name).join(', ');
    var exchng = new Exchange(dispatch, `consignment list ${manifest_names}`);

    getConsignments(manifests, exchng)
      .then(localSuccess)
      .catch(localError);
  };
};
