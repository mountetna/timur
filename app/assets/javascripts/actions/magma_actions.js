import {showMessages} from './message_actions';
import {Exchange} from './exchange_actions';
import * as MagmaAPI from '../api/magma_api';

export const addTemplate = (template)=>{
  return {
    type: 'ADD_TEMPLATE',
    model_name: template.name,
    template: template
  };
};

export const addDocumentsForTemplate = (model_name, documents)=>{
  return {
    type: 'ADD_DOCUMENTS',
    model_name: model_name,
    documents: documents
  };
};

export const reviseDocument = (document, template, attribute, revised_value)=>{
  return {
    type: 'REVISE_DOCUMENT',
    model_name: template.name,
    record_name: document[template.identifier],
    revision: {
      [attribute.name]: revised_value
    }
  }
};

export const discardRevision = (record_name, model_name)=>{
  return {
    type: 'DISCARD_REVISION',
    model_name: model_name,
    record_name: record_name
  };
};

export const addPredicates = (predicates)=>{
  return {
    type: 'ADD_PREDICATES',
    predicates
  };
};

/*
 * Here we add the models and documents to the store. At the same time we strip
 * off the model namespacing. The server returns the full name of the model. 
 * And we want that behavior so we can know from which project the data is
 * coming from. However, we do not want to propigate that namespacing to the UI.
 */
export const consumePayload = (dispatch, response)=>{
  if(response.models){
    Object.keys(response.models).forEach((model_name)=>{
      let model = response.models[model_name];

      if(model.template){
        dispatch(addTemplate(model.template));
      }

      if(model.documents){
        dispatch(addDocumentsForTemplate(model_name, model.documents));
      }
    });
  }
};

export const requestDocuments = (args)=>{
  let {
    model_name,
    record_names,
    attribute_names,
    filter,
    page,
    page_size,
    collapse_tables,
    exchange_name,
    success,
    error
  } = args;

  return (dispatch)=>{
    let localSuccess = (response)=>{

      if('error' in response){
        dispatch(showMessages([`There was a ${response.type} error.`]));
        console.log(response.error);
        return;
      }

      consumePayload(dispatch, response);
      if(success != undefined) success(response);
    };
  
    let localError = (e) => {
      if (!e.response) {
        dispatch(showMessages([`Something is amiss. ${e}`]));
        return;
      }
  
      e.response.json().then((response)=>{
        let errStr = response.errors.map((error)=> `* ${error}`);
        errStr = [`### Our request was refused.\n\n${errStr}`];
        dispatch(showMessages(errStr));
      });
  
      if(error != undefined){
        let message = JSON.parse(error.response);
        error(message);
      }
    };
  
    let get_doc_args = [
      {
        model_name,
        record_names,
        attribute_names,
        filter,
        page,
        page_size,
        collapse_tables
      },
      new Exchange(dispatch, exchange_name)
    ];
  
    MagmaAPI.getDocuments(...get_doc_args)
      .then(localSuccess)
      .catch(localError);
    }
};

export const requestModels = ()=>{
  let reqOpts = {
    'model_name': 'all',
    'record_names': [],
    'attribute_names': 'all',
    'exchange_name': 'request-models'
  };

  return requestDocuments(reqOpts);
};

export const requestIdentifiers = ()=>{
  let reqOpts = {
    'model_name': 'all',
    'record_names': 'all',
    'attribute_names': 'identifier',
    'exchange_name': 'request-identifiers'
  };

  return requestDocuments(reqOpts);
};

const revision_name = (model_name, record_name, attribute_name)=>(
  `revisions[${model_name}][${record_name}][${attribute_name}]`
);

// Append the revised data to a form.
const setFormData = (revisions, model_name)=>{
  var form = new FormData();
  for(var record_name in revisions){

    var revision = revisions[record_name];
    for(var attribute_name in revision){

      if(Array.isArray(revision[attribute_name])){

        revision[attribute_name].forEach((value)=>{
          let rev_nm = revision_name(model_name, record_name, attribute_name);
          return form.append(rev_nm+'[]', value);
        });
      }
      else{
        let rev_nm = revision_name(model_name, record_name, attribute_name);
        form.append(rev_nm, revision[attribute_name]);
      }
    }
  }

  return form;
};

export const sendRevisions = (model_name, revisions, success, error)=>{
  return (dispatch)=>{

    let localSuccess = (response)=>{
      consumePayload(dispatch, response);

      for(var record_name in revisions){
        dispatch(
          discardRevision(
            record_name,
            model_name
          )
        );
      }

      if(success != undefined) success();
    };

    let localError = (e)=>{
      e.response.json().then((response)=>{
        let errStr = response.errors.map((error)=> `* ${error}`);
        errStr = [`### The change we sought did not occur.\n\n${errStr}`];
        dispatch(showMessages(errStr));
      });

      if(error != undefined) error();
    };

    let exchng = new Exchange(dispatch, `revisions-${model_name}`);
    MagmaAPI.postRevisions(setFormData(revisions, model_name), exchng)
      .then(localSuccess)
      .catch(localError);
  }
};

// Download a TSV from magma via Timur.
export const requestTSV = (model_name,filter)=>{
  return (dispatch)=>{
    MagmaAPI.getTSVForm({model_name, filter, record_names: 'all'});
  };
};

export const requestAnswer = (question, callback)=>{
  return (dispatch)=>{

    let localSuccess = (response)=>{
      if('error' in response){
        dispatch(showMessages([`There was a ${response.type} error.`]));
        console.log(response.error);
        return;
      }

      if(callback != undefined) callback(response);
    };

    let localError = (error)=>{
      console.log(error);
    };

    let question_name = question;
    if(Array.isArray(question)){
      question_name = [].concat.apply([], question).join('-');
    }
    let exchange = new Exchange(dispatch, question_name);
    MagmaAPI.getAnswer(question, exchange)
      .then(localSuccess)
      .catch(localError);
  };
}

export const requestPredicates = ()=>{
  return (dispatch)=>{

    let localCallback = (response)=>{
      dispatch(addPredicates(response.predicates));
    };

    dispatch(requestAnswer('::predicates', localCallback));
  };
};
