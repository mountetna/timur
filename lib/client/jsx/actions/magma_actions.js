import {showMessages} from './message_actions';
import {Exchange} from './exchange_actions';
import {
  getDocuments,
  getAnswer,
  getTSVForm,
  postRevisions
} from '../api/magma_api';

export const ADD_TEMPLATE = 'ADD_TEMPLATE';
export const ADD_DOCUMENTS = 'ADD_DOCUMENTS';
export const REVISE_DOCUMENT = 'REVISE_DOCUMENT';
export const DISCARD_REVISION = 'DISCARD_REVISION';
export const ADD_PREDICATES = 'ADD_PREDICATES;';

export const addTemplate = (template) => {
  return {
    type: ADD_TEMPLATE,
    model_name: template.name,
    template: template
  };
};

export const addDocumentsForTemplate = (model_name, documents) => {
  return {
    type: ADD_DOCUMENTS,
    model_name: model_name,
    documents: documents
  };
};

export const reviseDocument = (
  document,
  template,
  attribute,
  revised_value
) => {
  console.log('in reviseDocument');
  return {
    type: REVISE_DOCUMENT,
    model_name: template.name,
    record_name: document[template.identifier],
    revision: {
      [attribute.name]: revised_value
    }
  };
};

export const discardRevision = (record_name, model_name) => {
  return {
    type: DISCARD_REVISION,
    model_name: model_name,
    record_name: record_name
  };
};

export const addPredicates = (predicates) => {
  return {
    type: ADD_PREDICATES,
    predicates
  };
};

/*
 * Here we add the models and documents to the store. At the same time we strip
 * off the model namespacing. The server returns the full name of the model.
 * And we want that behavior so we can know from which project the data is
 * coming from. However, we do not want to propigate that namespacing to the UI.
 */
export const consumePayload = (dispatch, response) => {
  if (response.models) {
    Object.keys(response.models).forEach((model_name) => {
      let model = response.models[model_name];

      if (model.template) {
        dispatch(addTemplate(model.template));
      }

      if (model.documents) {
        dispatch(addDocumentsForTemplate(model_name, model.documents));
      }
    });
  }
};

export const requestDocuments = (args) => {
  let {
    model_name,
    record_names,
    attribute_names,
    filter,
    page,
    page_size,
    collapse_tables,
    exchange_name
  } = args;

  return (dispatch) => {
    let handleRequestSuccess = (response) => {
      if ('error' in response) {
        dispatch(showMessages([`There was a ${response.type} error.`]));
        console.log(response.error);
        return;
      }

      consumePayload(dispatch, response);
      return Promise.resolve(response);
    };

    let handleRequestError = (e) => {
      if (!e.response) {
        dispatch(showMessages([`Something is amiss. ${e}`]));
        return;
      }

      e.response.json().then((response) => {
        let errStr = response.error
          ? response.error
          : response.errors.map((error) => `* ${error}`);
        errStr = [`### Our request was refused.\n\n${errStr}`];
        dispatch(showMessages(errStr));
      });
      return new Promise((resolve, reject) => {
        reject(e);
      });
    };

    let get_doc_args = [
      {
        model_name,
        record_names,
        attribute_names,
        filter,
        page,
        page_size,
        collapse_tables
      },
      new Exchange(dispatch, exchange_name)
    ];

    return getDocuments(...get_doc_args)
      .then(handleRequestSuccess)
      .catch(handleRequestError);
  };
};

export const requestModels = () => {
  let reqOpts = {
    model_name: 'all',
    record_names: [],
    attribute_names: 'all',
    exchange_name: 'request-models'
  };

  return requestDocuments(reqOpts);
};

export const requestIdentifiers = () => {
  let reqOpts = {
    model_name: 'all',
    record_names: 'all',
    attribute_names: 'identifier',
    exchange_name: 'request-identifiers'
  };

  return requestDocuments(reqOpts);
};

export const getRevisionTempUrl = (
  document,
  template,
  attribute,
  revised_value
) => {
  return (dispatch) => {
    const recordName = document[template.identifier];
    const revisions = {
      [recordName]: {
        [attribute.name]: revised_value
      }
    };

    let exchng = new Exchange(dispatch, `temp-url-revisions-${template.name}`);
    return postRevisions(formatRevisions(revisions, template.name), exchng);
  };
};

export const sendRevisions = (model_name, revisions, success, error) => {
  return (dispatch) => {
    let localSuccess = (response) => {
      consumePayload(dispatch, response);

      debugger;

      // Handle any file upload revisions by sending the file to Metis.
      // Here Magma's response should be a temporary upload URL.
      // 1. Grab the file from the original revisions list.
      // 2. Upload the file to Metis.
      // 3. Send another update to Magma with the file location to the
      //      record.
      // 4. Update the local store. (consumePayload?)

      for (var record_name in revisions) {
        dispatch(discardRevision(record_name, model_name));
      }

      if (success != undefined) success();
    };

    let localError = (e) => {
      e.response.json().then((response) => {
        let errStr = response.error
          ? response.error
          : response.errors.map((error) => `* ${error}`);
        errStr = [`### The change we sought did not occur.\n\n${errStr}`];
        dispatch(showMessages(errStr));
      });

      if (error != undefined) error();
    };

    let exchng = new Exchange(dispatch, `revisions-${model_name}`);
    postRevisions(formatRevisions(revisions, model_name), exchng)
      .then(localSuccess)
      .catch(localError);
  };
};

const formatRevisions = (revisions, model_name) => {
  const modelRevisions = {};
  modelRevisions[model_name] = revisions;

  const formattedRevs = {
    revisions: modelRevisions
  };
  return formattedRevs;
};

// Download a TSV from magma via Timur.
export const requestTSV = (model_name, filter, attribute_names) => {
  return (dispatch) => {
    getTSVForm(model_name, filter, attribute_names);
  };
};

export const requestAnswer = (question, callback) => {
  return (dispatch) => {
    let localSuccess = (response) => {
      if ('error' in response) {
        dispatch(showMessages([`There was a ${response.type} error.`]));
        console.log(response.error);
        return;
      }

      if (callback != undefined) callback(response);
    };

    let localError = (error) => {
      console.log(error);
    };

    let question_name = question;
    if (Array.isArray(question)) {
      question_name = [].concat.apply([], question).join('-');
    }
    let exchange = new Exchange(dispatch, question_name);
    getAnswer(question, exchange).then(localSuccess).catch(localError);
  };
};

export const requestPredicates = () => {
  return (dispatch) => {
    let localCallback = (response) => {
      dispatch(addPredicates(response.predicates));
    };

    dispatch(requestAnswer('::predicates', localCallback));
  };
};
