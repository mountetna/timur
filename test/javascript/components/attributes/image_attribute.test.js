import React from 'react';
import {mount, shallow} from 'enzyme';
import {mockStore} from '../../helpers';
import renderer from 'react-test-renderer';
import ButtonBar from '../../../../lib/client/jsx/components/button_bar';
import {STUB} from '../../../../lib/client/jsx/components/attributes/file_attribute';
import ImageAttribute from '../../../../lib/client/jsx/components/attributes/image_attribute';

import * as magmaActions from '../../../../lib/client/jsx/actions/magma_actions';

describe('ImageAttribute', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      location: {
        path: '/labors/browse/monster/Nemean Lion'
      }
    });
  });

  it('renders the button bar while editing', () => {
    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={null}
        value={null}
        mode='edit'
        attribute='gravatar'
        document='Timur'
        revised_value=''
        store={store}
      />
    );

    const buttons = component.find(ButtonBar);
    expect(buttons.exists()).toBeTruthy();

    const tree = renderer
      .create(
        <ImageAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={null}
          mode='edit'
          attribute='gravatar'
          document='Timur'
          revised_value=''
          store={store}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('does not render action buttons while not editing', () => {
    const value = {
      url: 'https://example.com?HMAC-header=foo'
    };

    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={null}
        value={value}
        mode='view'
        attribute='gravatar'
        document='Timur'
        revised_value=''
        store={store}
      />
    );

    const buttons = component.find('.file-buttons');
    expect(buttons.exists()).toBeFalsy();

    const thumbnail = component.find('.image-thumbnail');
    expect(thumbnail.exists()).toBeTruthy();

    const tree = renderer
      .create(
        <ImageAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
          store={store}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders thumbnail in view mode if file exists', () => {
    const value = {
      url: 'https://example.com?HMAC-header=foo'
    };

    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={null}
        value={value}
        mode='view'
        attribute='gravatar'
        document='Timur'
        revised_value=''
        store={store}
      />
    );

    const thumbnail = component.find('.image-thumbnail');
    expect(thumbnail.exists()).toBeTruthy();

    const tree = renderer
      .create(
        <ImageAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
          store={store}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders blank paths correctly while not editing', () => {
    let value = {path: '::blank'};

    let component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={null}
        value={value}
        mode='view'
        attribute='gravatar'
        document='Timur'
        revised_value=''
        store={store}
      />
    );

    expect(component.text().trim()).toEqual('Blank file');

    const tree = renderer
      .create(
        <ImageAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
          store={store}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders "missing file" correctly while not editing', () => {
    const value = null;

    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={null}
        value={value}
        mode='view'
        attribute='gravatar'
        document='Timur'
        revised_value=''
        store={store}
      />
    );

    expect(component.text().trim()).toEqual('No file');

    const tree = renderer
      .create(
        <ImageAttribute
          model_name='conquests'
          record_name='Persia'
          template={null}
          value={value}
          mode='view'
          attribute='gravatar'
          document='Timur'
          revised_value=''
          store={store}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('accepts a valid Metis path', () => {
    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={{name: 'Conquests', identifier: 1}}
        value={null}
        mode='edit'
        attribute={{name: 'ExpansionPlans'}}
        document={{'1': 'Timur'}}
        revised_value=''
        store={store}
      />
    );

    const cloudButton = component.find('.cloud').first();
    cloudButton.simulate('click');

    const metisPathInput = component.find('.file-metis-select').find('input');
    metisPathInput.instance().value = 'metis://project/bucket/file_name.txt';

    const checkButton = component.find('.check').first();
    checkButton.simulate('click');

    expect(component.find('file-metis-error').exists()).toBeFalsy();

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: 'metis://project/bucket/file_name.txt'}
        }
      }
    ]);
  });

  it('shows an error message when given an invalid Metis path', () => {
    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={{name: 'Conquests', identifier: 1}}
        value={null}
        mode='edit'
        attribute={{name: 'ExpansionPlans'}}
        document={{'1': 'Timur'}}
        revised_value=''
        store={store}
      />
    );

    const cloudButton = component.find('.cloud').first();
    cloudButton.simulate('click');

    const metisPathInput = component.find('.file-metis-select').find('input');
    metisPathInput.instance().value = 'project/bucket/file_name.txt';

    const checkButton = component.find('.check').first();
    checkButton.simulate('click');

    expect(component.find('.file-metis-error').exists()).toBeTruthy();

    const actions = store.getActions();
    expect(actions).toEqual([]);
  });

  it('dispatches an action to mark a file as blank', () => {
    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={{name: 'Conquests', identifier: 1}}
        value={null}
        mode='edit'
        attribute={{name: 'ExpansionPlans'}}
        document={{'1': 'Timur'}}
        revised_value=''
        store={store}
      />
    );

    const stubButton = component.find('.stub').first();
    stubButton.simulate('click');

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: STUB}
        }
      }
    ]);
  });

  it('dispatches an action to remove a file', () => {
    const component = mount(
      <ImageAttribute
        model_name='conquests'
        record_name='Persia'
        template={{name: 'Conquests', identifier: 1}}
        value={null}
        mode='edit'
        attribute={{name: 'ExpansionPlans'}}
        document={{'1': 'Timur'}}
        revised_value=''
        store={store}
      />
    );

    const removeButton = component.find('.remove').first();
    removeButton.simulate('click');

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: 'REVISE_DOCUMENT',
        model_name: 'Conquests',
        record_name: 'Timur',
        revision: {
          ['ExpansionPlans']: {path: null}
        }
      }
    ]);
  });

  // TODO: test for dispatching action when file uploaded
  //   * This has to talk to both Metis and Magma, so a little more
  //      involved.
});
