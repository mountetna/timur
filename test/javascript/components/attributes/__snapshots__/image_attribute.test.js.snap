// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageAttribute does not render action buttons while not editing 1`] = `
<div
  className="attribute"
>
   
  <a
    href="https://example.com?HMAC-header=foo"
  >
    <img
      className="image-thumbnail"
      src="https://example.com?HMAC-header=foo&thumbnail=true"
    />
  </a>
   
</div>
`;

exports[`ImageAttribute renders "missing file" correctly while not editing 1`] = `
<div
  className="attribute"
>
   
  <div
    className="document_empty"
  >
    No file
  </div>
   
</div>
`;

exports[`ImageAttribute renders blank paths correctly while not editing 1`] = `
<div
  className="attribute"
>
   
  <span
    className="file-blank"
  >
     Blank file 
  </span>
   
</div>
`;

exports[`ImageAttribute renders the button bar while editing 1`] = `
<div
  className="attribute file"
>
  <input
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="file-buttons"
  >
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Link a file from Metis"
    >
      <i
        className="icon cloud cloud fa fa-2x fa-cloud"
      />
    </span>
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Mark this file as blank"
    >
      <i
        className="icon stub meh-blank fa fa-2x fa-meh-blank"
      />
    </span>
    <span
      className="fa-stack fa-fw"
      onClick={[Function]}
      title="Remove this file link"
    >
      <i
        className="icon remove trash-alt fa fa-2x fa-trash-alt"
      />
    </span>
  </div>
  <span
    className="file-missing"
  >
     No file 
  </span>
</div>
`;

exports[`ImageAttribute renders thumbnail in view mode if file exists 1`] = `
<div
  className="attribute"
>
   
  <a
    href="https://example.com?HMAC-header=foo"
  >
    <img
      className="image-thumbnail"
      src="https://example.com?HMAC-header=foo&thumbnail=true"
    />
  </a>
   
</div>
`;
